### ISSUE TITLE
Scheduler Dies, IMAP import stops

### ISSUE THREAD

**Issue by dsukhin** [zammad/zammad#2043]
### Infos:

* Used Zammad version: 2.5.x (development branch)
* Installation method (source, package, ..): source
* Operating system: CentOS
* Database + version: MySQL
* Elasticsearch version: none
* Browser + version: N/A

### Expected behavior:

* IMAP job completes fully

### Actual behavior:

* Scheduler process dies without error messages

### Steps to reproduce the behavior:

* Attempt import IMAP mailbox with 27k messages
* tail -f log/production.log
* IMAP processing simply stops after a while (before completing)
* The system reports the scheduler PID is dead
* Restarting the scheduler makes the job restart and continue before dying again at a new place

Yes I'm sure this is a bug and no feature request or a general question.

This is likely related to #1861 #2013 and maybe #2016?

**Comment by thorsteneckel** [comment]
Are there any lines in `log/scheduler_err.log` or `log/scheduler_out.log`? What are the last lines in `log/production.log` when the job dies?

**Comment by dsukhin** [comment]
Unfortunately nothing in either of the scheduler logs files (they are there but empty). The last lines when the scheduler stops are usually actually unrelated to the imap job. Looks like its a call to the server in the background. This certainly isn't relevant when you look at the timing for the GET (almost 5 hours after the processing stopped).

```
I, [2018-05-23T10:58:18.820182 #24850]  INFO -- : 2018-05-23T10:58:18-0400: [Worker(host.-----.com pid:24850)] Job Observer::Ticket::UserTicketCounter::BackgroundJob (id=82661) COMPLETED after 0.0379
I, [2018-05-23T10:58:18.823219 #24850]  INFO -- : 2018-05-23T10:58:18-0400: [Worker(host:host.-----.com pid:24850)] Job Observer::Ticket::UserTicketCounter::BackgroundJob (id=82662) RUNNING
I, [2018-05-23T10:58:18.854454 #24850]  INFO -- : 2018-05-23T10:58:18-0400: [Worker(host:host.-----.com pid:24850)] Job Observer::Ticket::UserTicketCounter::BackgroundJob (id=82662) COMPLETED after 0.0310
I, [2018-05-23T10:58:18.859379 #24850]  INFO -- : 2018-05-23T10:58:18-0400: [Worker(host:host.-----.com pid:24850)] Job Transaction::BackgroundJob (id=82663) RUNNING
I, [2018-05-23T10:58:18.967361 #24850]  INFO -- : set_attributes_by_x_headers header x-zammad-ticket-customer_id found. Assign customer_id=8011
I, [2018-05-23T10:58:19.016268 #24850]  INFO -- : set_attributes_by_x_headers header x-zammad-article-preferences found. Assign preferences={"send-auto-response"=>true, "is-auto-response"=>false}
I, [2018-05-23T10:58:19.324266 #24850]  INFO -- :  - message 25898/27005 in INBOX
I, [2018-05-23T10:58:19.554926 #24850]  INFO -- : Process email with msgid '<1521678505@-----.com>'
I, [2018-05-23T15:33:28.616784 #24831]  INFO -- : Started GET "/" for ----- at 2018-05-23 15:33:28 -0400
I, [2018-05-23T15:33:28.699753 #24831]  INFO -- : Processing by InitController#index as HTML
I, [2018-05-23T15:33:28.801619 #24831]  INFO -- :   Rendering init/index.html.erb within layouts/application
I, [2018-05-23T15:33:28.802404 #24831]  INFO -- :   Rendered init/index.html.erb within layouts/application (0.6ms)
I, [2018-05-23T15:33:28.807101 #24831]  INFO -- : Completed 200 OK in 107ms (Views: 12.1ms | ActiveRecord: 34.5ms)
```

I have since tried to run the import while having the web interface fully closed. The job is completed now but did still randomly stop a few times. This time, the last line was even less informative, just up to `Process email with msgid '<...>'` part and then silence. Just like the above, but without the GET (which came later anyway). I may start another import on a new instance just to monitor it again and will provide more info if I can. So far, it's not clear what the root cause is and I can't find a single log with anything informative. Just a suddenly dead scheduler job. After restarting it, it runs again for a bit and stops again later. Very strange.

**Comment by thorsteneckel** [comment]
Thanks for your comprehensive analysis! Would be great if we would know how to reproduce it. From there on I can create debug script that will help us to find the root cause.

**Comment by dsukhin** [comment]
Reproducing is as simple as having a mailbox with A LOT of messages. 27,000 in my case. Should be easy enough to generate that mailbox of unique messages with a script. 

It's worth it to use the new "read only import" mode I made and rely on message ids to determine what was already imported so you can reuse the mailbox for tests (without things being marked as read in the process). Let me know if you have any questions on that. It's not a material change, just one that leaves the mailbox intact while processing.

**Comment by thorsteneckel** [comment]
Would be great if you could provide a system in which we can reproduce it. I will then provide a debugging script to find the cause 👍

**Comment by dsukhin** [comment]
Sure thing. I have a dev version set up that I can reset and we should see the issue reliably again. Let me know what I should do with the debug script when you provide it and when where I should put it in.

**Comment by dsukhin** [comment]
Also getting scheduler death when putting new rules in the schedule interface. Could be related.. Watching the logs as it happens, i see some "max_connection" errors to my mysql server in the background jobs. Dont think this would cause the scheduler to die but still might be relevant.

I can raise the limit certainly but is there a similar setting in zammad to limit the number of sql connections to a reasonable level?

**Comment by dsukhin** [comment]
Comment above (re: pool) is unrelated to the scheduler stopping issue during import. Awaiting debug scripts and will test further.

**Comment by dsukhin** [comment]
Any update on the debug scripts? I'm importing again and scheduling dies pretty frequently.

**Comment by thorsteneckel** [comment]
Hi @dsukhin - sorry for the late reply. I created a little script that is basically the Scheduler but instead of running in a background job it will run in the foreground. So the script will run as long as the Scheduler is not dying (You can run `tail -f debug_issue.log` to see what's currently going on). Please don't run this and the regular Scheduler at the same time. Please do as follows:

1.) Create a file called `debug_issue.rb` in your Zammad directory (usually `/opt/zammad`)
2.) Run the file from your Zammad directory via `zammad run rails r debug_issue.rb` or `rails r debug_issue.rb` as zammad user, depending on your installation source (package/source)
3.)A `debug_issue.log` file will be created. Post the file here. Make sure all sensitive data is anonymized! If you want you can drop it as an email to support@zammad.com. Please refer to this issue and me :)
4.) Delete the `debug_issue.rb` file

```ruby
filename = File.basename($0, '.rb')

# redirect logging
Rails.logger = Logger.new("#{filename}.log").tap do |logger|
  logger.level = :debug
end
ActiveRecord::Base.logger = Rails.logger.clone

Scheduler.threads
```

Thanks for your support!

**Comment by dsukhin** [comment]
Thanks @thorsteneckel. I ran this 2 times as `rails r debug_issue.rb &> issue_log.log &` to capture both the debug log and any stdout/err. Nothing appeared in `issue_log.log` except some unprocessable mails and charset errors that are properly caught. The `debug_issue.log` created by your script is also less than informative -- both logs end at different points (different things being executed) and no errors are logged. Running again right now to see what pops up... not sure what's happening tho. Once I have 3 debug logs, I'll send them via email to see if you notice anything more than I have.

**Comment by dsukhin** [comment]
Figured out what is going on here. It seems that during a large IMAP import, memory continues to be collected until the OS kills the job. In other words, it's a memory leak, perhaps in the imap driver or email processing logic. Will dive deeper but would love any advice on how to debug this most effectively.

**Comment by thorsteneckel** [comment]
Thanks for the update! That sounds like a reasonable assumption. We already did some memory leak debugging in the past but it always was kind of cumbersome. Let me discuss this with the others and I'll come back to you.

**Comment by martini** [comment]
Update: @thorsteneckel I can reproduce in with Zammad 2.5. If you need more infos, let me know.

**Comment by thorsteneckel** [comment]
Hi @dsukhin,

I just tried to reproduce this with an inbox of ~14k entries. I did various memory profiling sessions and watched the memory consumption closely. I saw a tiny "leak" for about 1MB per 100 processed mails. Which means that would be 270MB for your 27k inbox. This "leak" is no real leak because it's caused by the instantiation of a lot of string for character encoding checks.

The question why our test results differ remains. I have three ideas of what this may caused by:

1. My processed mails differ in a relevant way with the ones of you. I generated mine this way:
```ruby
27_000.times do |count|

  puts count

  Mail.deliver do
    from         'some@mail.com'
    to           'some@mail.com'
    subject      "This is test email no. #{count}"
    content_type 'text/plain; charset=UTF-8'
    body         OpenSSL::Random.random_bytes(4000).to_s
  end
end
```

2. I tested from the wrong context. In my earlier script I executed all Scheduler jobs. In my current test script I only fetch from my IMAP server (run it via `rails r *filename*.rb`:
```ruby
options = {
  host:     'host.example.com',
  port:     993,
  ssl:      true,
  user:     'some_user',
  password: 'some_password',
  folder:   'INBOX'
}

Channel::Driver::Imap.new.fetch(options, Channel.last)
```

3. Your [local changes](https://github.com/zammad/zammad/pull/1998) introduced this error. Could you please run the tests in both envs: Your customized codebase and a blank `develop` state?

**Comment by thorsteneckel** [comment]
PS: @martini improved the longevity of scheduler processes a lot by [this commit](https://github.com/zammad/zammad/commit/cc4ce9207277c82c4e645cfc2a528c18cd93d551#diff-a1dd81206f9f855477e929464e1e638b). However, this has no relation to the described memory leak.

**Comment by dsukhin** [comment]
I will give the latest develop branch another test. I do recall testing originally on a clean branch and still had the issue.

From what you mentioned above - I also have a few ideas: 

1. I have a number of emails much larger than 4kb in size... perhaps too many of those in succession causes it? You mention a "leak" of 1mb per 100 messages of 4kb in size. My total mailbox size is ~750mb making my average message 8x the size of your test message - in reality its a small number of large messages and mostly smaller ones. If the character encoding checks leak scales with message size (which i presume it might so that it can parse the whole message body) that would mean a 1mb "leak" every ~12 messages which can quickly get out of hand on a busy server. 
 
2. Not sure if anything else isn't being run when you only test the import script... does that do all the parsing and insert into all the dbs? May be worth it to test fully.

**Comment by thorsteneckel** [comment]
Hi @dsukhin - thanks for the reply. I'll retry my test with larger mails. Quick question: How much memory do you have available and with how much memory does the process start/die?

**Comment by MrGeneration** [comment]
@dsukhin is this topic still up to date or did the problem disappear with the last versions?

**Comment by dsukhin** [comment]
@MrGeneration, i am still on a forked version of 2.5 since the two way sync PR hasn't had any movement and its critical for my organization. I have been meaning to merge back the changes to keep the PR fresh - would love to understand more about why its not considered a mergeable feature even tho its fully optional and useful for many orgs/scenarios if only to keep a data backup on an open protool.

I do continue to have trouble with the scheduler crashing and updates not running. I will have to merge back to 2.10 to let you know if this is gone in the new versions.

**Comment by saz** [comment]
#2486 might be related, as scheduler was dying for me without a useful error message on such mails.

**Comment by MrGeneration** [comment]
@saz could you please update to the latest Zammad version and watch this?
Because issue #2486 has been resolved and pushed to stable and thus should be available already.

**Comment by saz** [comment]
@MrGeneration I'm on 2.9.0 and as far as I can see 2.9.1 isn't released yet and #2486 is marked for that milestone.
But replacing all/models/store.rb with the current dev version resolves the issue for me.

**Comment by MrGeneration** [comment]
Thanks for the Feedback.
I will close this issue for now, as I expect this to be fixed at the moment (sorry, I ignore forks and not up to date versions here).

As for the two-way mail sync PR I can't say too much. I think Thorsten stated the current reason for this.
I'm not in that part too deep, sorry!
